### ISSUE TITLE
C-Plugin: sensors use wrong rules repository keys

### ISSUE THREAD

**Issue by chrbayer** [SonarOpenCommunity/sonar-cxx#1521]
### Description

For a C test project I used the python version of scan-build command to generate clang analyzer reports with the command as described in the Wiki:

scan-build.py -plist --intercept-first --analyze-headers -o source-check/analyzer_reports make

The reports are generated as expected.

Using sonar-scanner with the property sonar.c.clangsa.reportPath=source-check/analyzer_reports/*/*.plist set in sonar-project.properties the reports are completely ignored, 
even with -X no clue is shown.

I can provide the report files if needed.

### Related information

* cxx plugin version 1.1.0
* SonarQube version 7.2.1

**Comment by asunnys** [comment]
请问基于clang的c代码分析器怎么集成到sonarqube的,能提供一下思路么

**Comment by ivangalkin** [comment]
Hi @chrbayer,

`sonar.c.clangsa.reportPath=<path>` has to point to the existing file. It's not a wildcard. Please try to set an absolute path or relative path to your project base dir.

Please also keep in mind, that clang static analyzer version >= 3.7 is supported only (#1442).

If the problem persists please provide the debug log. Thank you in advance.

**Comment by chrbayer** [comment]
Hi @ivangalkin,

I'm using the latest clang release: clang version 6.0.1 (tags/RELEASE_601/final)

In the wiki page using wildcards the way I do is explicit allowed and is needed, because clang static analyzer created one result file per source file.

~~~~
sonar.cxx.clangsa.reportPath
Scope: project 
Default: ---

Ant pattern describing the path to Clang Static Analyzer plist reports. Path can be relative or absolute. Single path or comma separated list of paths is supported. 

Examples:
sonar.cxx.clangsa.reportPath=divzero.plist
or
sonar.cxx.clangsa.reportPath=analyzer_reports/*/*.plist
~~~~

But I tested one single result file without wildcards, too. It was the same effect...

Please find attached one of the generated plist reports.

Thank you very much in advance!
[report-c3mbo998.plist.txt](https://github.com/SonarOpenCommunity/sonar-cxx/files/2218846/report-c3mbo998.plist.txt)

**Comment by ivangalkin** [comment]
HI @chrbayer,

> In the wiki page using wildcards the way I do is explicit allowed and is needed, because clang static analyzer created one result file per source file

Sorry, I was wrong w.r.t. wildcards.

Please provide the debug (`-X`) output too and the properties file with the project's settings if possible.

**Comment by ivangalkin** [comment]
@chrbayer one more question: Is it your initial SonarQube setup with Sonar-CXX? or do you experinece the problem after plugin update? Have you activated the corresponding rules in the Web-UI?

![clang-sa_repository](https://user-images.githubusercontent.com/11503715/43075074-c69ba4f6-8e7f-11e8-83e4-8c9f62a86cd7.png)

**Comment by chrbayer** [comment]
@ivangalkin 

It is my initial setup, so I have no experience with older versions of the plugin. I confirmed that I have the same number 111 of available rules for Clang-SA.

I have attached my properties file and the debug log, but there is nothing mentioned in the log regarding clangsa... Looks like it is completely ignored...

Thanks in advance!
[sonar-project.properties.txt](https://github.com/SonarOpenCommunity/sonar-cxx/files/2219624/sonar-project.properties.txt)
[log.txt](https://github.com/SonarOpenCommunity/sonar-cxx/files/2219626/log.txt)

**Comment by ivangalkin** [comment]
Hi @chrbayer,

the number of 111 is just the number of existing rules; you have to activate the Clang-SA rules in order to activate the Clang-SA [importing] sensor. I kindly ask you to read my answer in #1518 especially the block **Why you don't see any results (2)**: you have to create your own quality profile and activate the Clang-SA rules.

The fact, that the rules have to be activated before analysis is missing in our wiki. We will update the documentation.

**Comment by chrbayer** [comment]
Hi @ivangalkin,

so yes, the rules for Clang-SA have not been activated. I was surprised that the rules for cppcheck e.g. are active and others not...

I created a new Quality Profile, activated the rules for Clang-SA, setup the project to use the new Quality Profile, but I still do not see output from a new sensor, only SquidSensor [c], RatsSensors [c] and CppCheckSensor [c]...

**Comment by ivangalkin** [comment]
>  I was surprised that the rules for cppcheck e.g. are active and others not...

not sure, but this might have some backwards-compatibility reasons; maybe @guwirth knows if it's a bug or a feature ;-)

W.r.t. to the still missing Clang-SA warnings: Could you please re-run your analysis with `-X` and provide the logs? Sorry for the inconvenience.

**Comment by chrbayer** [comment]
I double checked, the the new Profile "PAF way" has all 111 Clang-SA rules activated and that it is used by the PAF project...

Please find the logs attached.

Thank you very mich for your help!

**Comment by ivangalkin** [comment]
your log still contains the entry

```
17:54:43.243 DEBUG: 'C (Community) ClangSASensor' skipped because there is no related rule activated in the quality profile
```

before I'll dig deeper, could you please double-check that you've activated the rules for the language **C (Community)** and the updated the quality profile for the **C (Community)**

As far as I see you use both C and C++ plugins in parallel. They have their separate quality profiles and separate rules (although de-facto the scope is the same). So it's easy to mix the settings.

By the way, please consider to use the C++ plugin only. De-facto it's the same plugin as the C one. It is capable to import results for C-language too. You'll have to 1) deinstall the sonar-c-plugin 2) adapt the properties...

```
sonar.cxx.suffixes.sources
sonar.cxx.suffixes.headers
sonar.cxx.cFilesPatterns
```
... or just to reset them to default.

**Comment by ivangalkin** [comment]
Hi @chrbayer! I'm glad, that you got the plugin working. You recognized correctly, that the benefit of sonar-c-plugin is questionable.

@guwirth I believe I've found the problem. Our language dependent sensor definitions (this time it's the rules repository key) are inconsistent. I'll submit the fix, but I ask myself, if we have some general design problem. I don't think that the sensors (their rules) must exist twice. SonarCFamily manages to combine two language in one plugin. Why can't we do the same.

**Comment by jmecosta** [comment]
i think as people have mention earlier, besides the metrics you are able to
define tailored profiles for c and c++ witch is a very good thing if you
think about both style and new c++ languages features that can cause to
many noise in sonar.

i would rather avoid either poluting the source code with ignore tags or
having to handle those in the web ui.

having dedicated profiles for them its really usefull (headers we need to
be cautious tought).

**Comment by jmecosta** [comment]
@guwirth it should be ok to merge the all thing in one jar file. but we still need to define two languages in one plugin and in the end of the day is just merging complexity into one plugin because the end result in terms of functionality its pretty much the same. end user will still see two languages, two profiles, etc. 

after this, if we have two languages in one plugin the sensors will need to have some logic to detect what language is the resource before picking the profile. As it is now, its quite simple since the all code is all the same.

But im more than fine if you decide to join them, however its important to keep the possibility to tailor the profiles per language

**Comment by guwirth** [comment]
>> I was surprised that the rules for cppcheck e.g. are active and others not...

> not sure, but this might have some backwards-compatibility reasons; maybe @guwirth knows if it's a bug or a feature ;-)

@ivangalkin think at the beginning there was no other sensor for external tools. After adding more and more sensors the whole plugin was getting slower and slower. Therefore decision was to disable the others. Meanwhile there are a lot of ways for static code analysis and cppcheck seems to be no more the default.
* we can enable all rules per default
  * advantage: easy to use
  * disadvantage: slow per default (or maybe no more with SQ 6.7 LTS++?)
* we can disable all rules per default
  * advantage: you have to enable the rules you like to use - pay only what you are using
  * disadvantage: no rule support out of the box

**Comment by ivangalkin** [comment]
@guwirth I like the disabling all rules per default as a consistent solution with low performance impact.

**Comment by guwirth** [comment]
> after this, if we have two languages in one plugin the sensors will need to have some logic to detect what language is the resource before picking the profile. As it is now, its quite simple since the all code is all the same.

@jmecosta not sure if this is the case?

language1
   => extension1
   => repository1
language2
   => extension2
   => repository2

report => filename =>  extension => language => repository

**Comment by jmecosta** [comment]
Exactly you have to put that logic there. Will it make the sensor slower?
Maybe its neglectible.

One think i still dont confused. How will files get indexed?

**Comment by jmecosta** [comment]
Nevermin the indexing. Its defined in the language.

**Comment by jmecosta** [comment]
One last thing from me. What is exactly the problem we are solving by
merging them?

Im a bit confused since there are quite a few issues with c, c++, paid c++,
why both c and cxx, etc and dont see how the merge would fix them

But for sure this issue is not related with any of those issues.

**Comment by asunnys** [comment]
@ivangalkin hi,我把cppcheck集成到里边可以用了,但是我还是不懂其中的原理,sonar-c-plugin本身是集成c代码静态检查的么 还是必须需要cppcheck这样的工具检查结果

**Comment by ivangalkin** [comment]
Hi @asunnys 

(I am a bit confused, that we communicate through google-translate, but why not 😊)

sonar-c-plugin (as well as sonar-cxx-plugin) have both:

a) an integrated static analyzer with a very limited number of checks
b) sensors, which can import analysis reports from 3rd party tools

For (a) it's enough to activate corresponding rules. Please see https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Activate-quality-rules. The relevant rules set is called "c[++] SonarQube" aka "SquidSensor".

For (b) you have to

1. run the tool first (e.g. cppcheck)
2. parametrize your sonar-scanner (or sonar maven plugin/ sonar gradle plugin) with corresponding paths to the tool's report (e.g. cppcheck produces an XML file)
3. run sonar-scanner; the importing sensor will parse the input file and integrate the issue into SonarQube

The detailed information can be found in https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Running-tools. Don't forget to activate the rule set for the required sensor, too (https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Activate-quality-rules).

You might want to read https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Running-the-analysis
Or even more detailed official SonarQube documentation https://docs.sonarqube.org/display/SONAR/Analyzing+Source+Code

If you miss something in the documentation, please open a separate issue. This one is already closed.

Regards

**Comment by asunnys** [comment]
Hi@ivangalkin 首先非常感谢您提供的帮助,它已经可以完美运行了,在过去一段时间因为非常忙,没有能即时回顾您,为此感到抱歉

**Comment by ivangalkin** [comment]
Hi @asunnys! I’m glad the plugin works well. No problem about the late reply. Feel free to open an issue if you will have some questions/problems. Regards.
